rust/agama-server/src/web/common/issues.rs

```diff
@@ -0,0 +1,280 @@
+    /// * Relevant events from D-Bus.
+    pub async fn start(dbus: zbus::Connection, events: EventsSender) -> IssuesClient {
+        let (tx, rx) = mpsc::channel(16);
+        let mut service = IssuesService {
+            issues: HashMap::new(),
+            dbus,
+            events,
+            commands: rx,
+        };
+
+        tokio::spawn(async move { service.run().await });
+        IssuesClient(tx)
+    }
+
+    /// Main loop of the service.
+    async fn run(&mut self) {
+        let mut messages = build_properties_changed_stream(&self.dbus).await.unwrap();
+        loop {
+            tokio::select! {
+                Some(cmd) = self.commands.recv() => {
+                    if let Err(e) = self.handle_command(cmd).await {
+                        tracing::error!("{e}");
+                    }
+                }
+
+                Some(Ok(message)) = messages.next() => {
+                    if let Some(changed) = PropertiesChanged::from_message(message) {
+                        if let Err(e) = self.handle_property_changed(changed) {
+                            tracing::error!("IssuesService: could not handle change: {:?}", e);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Handles commands from the client.
+    async fn handle_command(&mut self, command: IssuesCommand) -> IssuesServiceResult<()> {
+        match command {
+            IssuesCommand::Get(service, path, tx) => {
+                let issues = self.get(&service, &path).await?;
+                tx.send(issues)
+                    .map_err(|_| IssuesServiceError::SendIssues)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Handles PropertiesChanged events.
+    fn handle_property_changed(&mut self, message: PropertiesChanged) -> IssuesServiceResult<()> {
+        let args = message.args()?;
+        let inner = message.message();
+        let header = inner.header();
+        let Some(path) = header.path() else {
+            return Ok(());
+        };
+
+        // Only process Agama issues signals.
+        if args.interface_name.as_str() != "org.opensuse.Agama1.Issues" {
+            return Ok(());
+        }
+
+        let Some(all) = args.changed_properties().get("All") else {
+            return Ok(());
+        };
+
+        let all = all.downcast_ref::<&Array>()?;
+        let issues = all
+            .into_iter()
+            .map(Issue::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
+
+        self.issues.insert(path.to_string(), issues.clone());
+
+        let event = Event::IssuesChanged {
+            path: path.to_string(),
+            issues,
+        };
+        self.events.send(event)?;
+        Ok(())
+    }
+
+    /// Gets the issues for a given D-Bus service and path.
+    async fn get(&mut self, service: &str, path: &str) -> IssuesServiceResult<Vec<Issue>> {
+        if let Some(issues) = self.issues.get(path) {
```

Comment:
so this is basically that cache implementation. Maybe it should be mentioned in comment?